hazelcast/src/main/java/com/hazelcast/sql/impl/client/ReadOptimizedLruCache.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.client;
+
+import java.util.Comparator;
+import java.util.PriorityQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+/**
+ * Implementation of an LRU cache optimized for read-heavy use cases.
+ * <p>
+ * It stores the entries in a {@link ConcurrentHashMap}, along with the last
+ * access time. It allows the size to grow beyond the capacity, up to
+ * `cleanupThreshold`, at which point the inserting thread will remove a batch
+ * of the eldest items in two passes.
+ * <p>
+ * The cleanup process isn't synchronized to guarantee that the capacity is not
+ * exceeded. The cache is available during the cleanup for reads and writes. If
```

Comment:
From my understanding, one way is that `put` only checks for weakly consistent size, so it might happen that multiple simultaneous put calls pass the size() gate and proceed to increase the cache size. Of course any put that will see the real size after some time will start the clean up so it all probably "evens out". Either way I haven't see the cache exceed the threshold in my perf tests.